File: python/atlasent/governance_agents.py
"""Constrained governance agents -- read-side SDK surface.

Three endpoints, all GET, all org-scoped server-side:
    GET /v1/governance/agents
    GET /v1/governance/findings?change_id=...
    GET /v1/governance/evaluations?change_id=...

``can_authorize`` is ``False`` on every agent and finding -- enforced
by a CHECK constraint on the runtime DB, not just convention.
Governance agents are advisory-only; no invocation endpoint is
exposed here. CI invocation is the responsibility of atlasent-action.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

_SEVERITY_RANK: dict[str, int] = {
    "info": 1,
    "low": 2,
    "medium": 3,
    "high": 4,
    "blocker": 5,
}


class AgentEvidenceRef(BaseModel):
    """A reference to an evidence artifact produced by a governance agent."""

    kind: str
    ref: str
    note: str | None = None

    model_config = ConfigDict(extra="allow", populate_by_name=True)


class GovernanceAgent(BaseModel):
    """An advisory governance agent registered in the AtlaSent registry.

    ``can_authorize`` is always ``False`` -- a structural invariant
    enforced by the runtime DB. Advisory agents produce findings;
    they never satisfy a permit gate.
    """

    slug: str
    version: str
    name: str
    description: str
    applicable_subject_kinds: list[str] = Field(default_factory=list)
    authority_class: Literal["advisory"]
    can_authorize: Literal[False]
    capabilities: list[str] = Field(default_factory=list)
    is_active: bool
    created_at: str
    retired_at: str | None = None

    model_config = ConfigDict(extra="allow", populate_by_name=True)


class GovernanceAgentFinding(BaseModel):
    """A finding emitted by a governance agent evaluation run.

    ``can_authorize`` is always ``False`` -- a DB CHECK constraint
    prevents any finding from satisfying a permit gate.
    """

    id: str
    org_id: str
    evaluation_id: str
    change_id: str
    agent_slug: str
    agent_version: str
    finding_type: str
    severity: str
    confidence: float | None = None
    summary: str
    evidence_refs: list[AgentEvidenceRef] = Field(default_factory=list)
    required_authority: str | None = None
    recommended_action: str | None = None
    can_authorize: Literal[False]
    supersedes_finding_id: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: str
    routed_gate_id: str | None = None

    model_config = ConfigDict(extra="allow", populate_by_name=True)


class GovernanceAgentEvaluation(BaseModel):
    """A governance agent run record for a single governed change."""

    id: str
    org_id: str
    change_id: str
    agent_slug: str
    agent_version: str
    input_hash: str
    status: str
    highest_severity: str | None = None
    findings_count: int
    summary: str | None = None
    runtime_ms: int | None = None
    failure_reason: str | None = None
    invoked_by_kind: str
    invoked_by: str | None = None
    started_at: str
    completed_at: str | None = None

    model_config = ConfigDict(extra="allow", populate_by_name=True)


@dataclass
class ListGovernanceAgentsResult:
    """Return type for :meth:`~atlasent.AtlaSentClient.list_governance_agents`."""

    agents: list[GovernanceAgent] = field(default_factory=list)
    rate_limit: Any | None = None


@dataclass
class ListGovernanceFindingsResult:
    """Return type for :meth:`~atlasent.AtlaSentClient.list_governance_findings`."""

    findings: list[GovernanceAgentFinding] = field(default_factory=list)
    rate_limit: Any | None = None


@dataclass
class ListGovernanceEvaluationsResult:
    """Return type for :meth:`~atlasent.AtlaSentClient.list_governance_evaluations`."""

    evaluations: list[GovernanceAgentEvaluation] = field(default_factory=list)
    rate_limit: Any | None = None


def highest_agent_finding_severity(
    findings: list[GovernanceAgentFinding],
) -> str | None:
    """Return the highest severity string across a list of findings.

    Returns ``None`` when the list is empty or all severities are
    unrecognized. Severity rank: ``blocker`` > ``high`` > ``medium`` >
    ``low`` > ``info``.
    """
    best: str | None = None
    rank = 0
    for f in findings:
        r = _SEVERITY_RANK.get(f.severity, 0)
        if r > rank:
            rank = r
            best = f.severity
    return best

File: typescript/src/governanceAgents.ts
/**
 * Constrained governance agents — read-side SDK surface.
 *
 * Three endpoints, all GET, all org-scoped server-side:
 *
 *   GET /v1/governance/agents                — registry of advisory agents
 *   GET /v1/governance/findings?change_id=…  — findings against one change
 *   GET /v1/governance/evaluations?change_id=… — agent run records
 *
 * **Doctrine — evaluation ≠ authorization ≠ execution.**
 * Every type in this module is read-only signal. `can_authorize` is
 * pinned `false` on the wire (DB-generated column on the registry,
 * CHECK on findings). The SDK does not expose an invocation method:
 * agent invocation is a CI concern (atlasent-action `governance-agents`
 * mode), not an application concern. This module is for surfaces that
 * want to render findings alongside the authority workflow.
 *
 * Wire schema source of truth lives in
 *   atlasent-api/packages/types/src/governance-agents.ts
 * which is intentionally standalone (not re-exported from @atlasent/types).
 * The shapes mirrored below are the read-side subset.
 *
 * @module
 */

// ─── enums (mirror SQL CHECK domains) ────────────────────────────────────────

export type AgentFindingSeverity =
  | "info"
  | "low"
  | "medium"
  | "high"
  | "blocker";

export type AgentEvaluationStatus =
  | "running"
  | "completed"
  | "failed"
  | "timeout";

export type AgentAuthorityDomain =
  | "engineering"
  | "runtime_platform"
  | "security"
  | "compliance"
  | "release_management"
  | "operations"
  | "customer_impact"
  | "governance_office";

export type AgentInvokerKind =
  | "human"
  | "service_account"
  | "autonomous_agent"
  | "system";

export type AgentSubjectKind =
  | "pull_request"
  | "schema_migration"
  | "runtime_flag"
  | "deployment"
  | "operational_rollout"
  | "regulated_execution_change"
  | "policy_bundle";

// ─── records ─────────────────────────────────────────────────────────────────

/**
 * A versioned advisory agent definition. `authority_class` is fixed to
 * `advisory` and `can_authorize` to `false` at the schema level — these
 * cannot be relaxed without a structural change to the runtime DB.
 */
export interface GovernanceAgent {
  readonly slug: string;
  readonly version: string;
  readonly name: string;
  readonly description: string;
  readonly applicable_subject_kinds: readonly AgentSubjectKind[];
  readonly authority_class: "advisory";
  /** Structurally false. Generated column on the runtime DB. */
  readonly can_authorize: false;
  readonly capabilities: readonly string[];
  readonly is_active: boolean;
  readonly created_at: string;
  readonly retired_at: string | null;
}

/** A typed evidence pointer attached to a finding. Free-form by design. */
export interface AgentEvidenceRef {
  readonly kind: string;
  readonly ref: string;
  readonly note?: string;
}

/**
 * One advisory finding produced by an agent run. `can_authorize` is
 * pinned `false` by a CHECK constraint on the underlying table — no
 * finding row in any environment can ever satisfy a gate.
 */
export interface GovernanceAgentFinding {
  readonly id: string;
  readonly org_id: string;
  readonly evaluation_id: string;
  readonly change_id: string;
  readonly agent_slug: string;
  readonly agent_version: string;
  readonly finding_type: string;
  readonly severity: AgentFindingSeverity;
  readonly confidence: number | null;
  readonly summary: string;
  readonly evidence_refs: readonly AgentEvidenceRef[];
  readonly required_authority: AgentAuthorityDomain | null;
  readonly recommended_action: string | null;
  /** Structurally false. CHECK constraint on the runtime DB. */
  readonly can_authorize: false;
  readonly supersedes_finding_id: string | null;
  readonly payload: Readonly<Record<string, unknown>>;
  readonly created_at: string;
  /**
   * Populated by the finding→gate routing trigger (atlasent-api #842).
   * Null when no matching gate exists at insertion time; can be
   * back-resolved by `governance_resolve_findings_for_gate(gate_id)`.
   */
  readonly routed_gate_id?: string | null;
}

/**
 * An append-only record of one agent run against one governed change.
 * The same (agent_slug, agent_version, input_hash) combination may
 * produce multiple rows across time — the runtime DB does not dedupe.
 */
export interface GovernanceAgentEvaluation {
  readonly id: string;
  readonly org_id: string;
  readonly change_id: string;
  readonly agent_slug: string;
  readonly agent_version: string;
  readonly input_hash: string;
  readonly status: AgentEvaluationStatus;
  readonly highest_severity: AgentFindingSeverity | null;
  readonly findings_count: number;
  readonly summary: string | null;
  readonly runtime_ms: number | null;
  readonly failure_reason: string | null;
  readonly invoked_by_kind: AgentInvokerKind;
  readonly invoked_by: string | null;
  readonly started_at: string;
  readonly completed_at: string | null;
}

// ─── response envelopes ──────────────────────────────────────────────────────

export interface ListGovernanceAgentsResponse {
  readonly agents: readonly GovernanceAgent[];
}

export interface ListGovernanceFindingsResponse {
  readonly findings: readonly GovernanceAgentFinding[];
}

export interface ListGovernanceEvaluationsResponse {
  readonly evaluations: readonly GovernanceAgentEvaluation[];
}

// ─── query shapes ────────────────────────────────────────────────────────────

export interface ListGovernanceFindingsQuery {
  readonly change_id: string;
  /** Optional: filter to one agent's findings. */
  readonly agent_slug?: string;
}

export interface ListGovernanceEvaluationsQuery {
  readonly change_id: string;
  /** Optional: filter to one agent's runs. */
  readonly agent_slug?: string;
}

// ─── helpers ─────────────────────────────────────────────────────────────────

const SEVERITY_RANK: Record<AgentFindingSeverity, number> = {
  info: 1,
  low: 2,
  medium: 3,
  high: 4,
  blocker: 5,
};

/**
 * Return the worst severity across a set of findings, or `null` when
 * the input is empty. Pure function, exported because every consumer
 * needs the same rollup logic (Console finds panel, CI summary, etc.).
 */
export function highestAgentFindingSeverity(
  findings: readonly Pick<GovernanceAgentFinding, "severity">[],
): AgentFindingSeverity | null {
  let best: AgentFindingSeverity | null = null;
  let rank = 0;
  for (const f of findings) {
    const r = SEVERITY_RANK[f.severity];
    if (r > rank) {
      rank = r;
      best = f.severity;
    }
  }
  return best;
}
